src/librustc_mir/shim.rs

```diff
@@ -259,6 +272,164 @@ impl<'a, 'tcx> DropElaborator<'a, 'tcx> for DropShimElaborator<'a, 'tcx> {
     }
 }
 
+/// Build a `Clone::clone` shim for `recvr_ty`. Here, `def_id` is `Clone::clone`.
+fn build_clone_shim<'a, 'tcx>(tcx: ty::TyCtxt<'a, 'tcx, 'tcx>,
+                              def_id: DefId,
+                              rcvr_ty: ty::Ty<'tcx>)
+                              -> Mir<'tcx>
+{
+    let sig = tcx.fn_sig(def_id);
+    let sig = tcx.erase_late_bound_regions(&sig);
+    let span = tcx.def_span(def_id);
+
+    debug!("build_clone_shim(def_id={:?})", def_id);
+
+    let mut local_decls = local_decls_for_sig(&sig, span);
+    let source_info = SourceInfo { span, scope: ARGUMENT_VISIBILITY_SCOPE };
+
+    let mut blocks = IndexVec::new();
+    let block = |blocks: &mut IndexVec<_, _>, statement, kind| {
+        blocks.push(BasicBlockData {
+            statements: vec![statement],
+            terminator: Some(Terminator { source_info, kind }),
+            is_cleanup: false
+        })
+    };
+
+    let rcvr = Lvalue::Local(Local::new(1+0)).deref();
+
+    let call_clone = |i, ty, rcvr_field, blocks: &mut _, local_decls: &mut IndexVec<_, _>| {
+        // `func == Clone::clone(&ty) -> ty`
+        let substs = Substs::for_item(tcx, def_id, |_, _| tcx.types.re_erased, |_, _| ty);
+        let func = Operand::Constant(box Constant {
+            span: span,
+            ty: tcx.mk_fn_def(def_id, substs),
+            literal: Literal::Value {
+                value: ConstVal::Function(def_id, substs),
+            },
+        });
+
+        let ref_loc = Lvalue::Local(
+            local_decls.push(temp_decl(
+                Mutability::Not,
+                tcx.mk_ref(tcx.types.re_erased, ty::TypeAndMut {
+                    ty,
+                    mutbl: hir::Mutability::MutImmutable,
+                }),
+                span
+            ))
+        );
+
+        let loc = Lvalue::Local(
+            local_decls.push(temp_decl(
+                Mutability::Not,
+                ty,
+                span
+            ))
+        );
+
+        // `let ref_loc: &ty = &rcvr_field;`
+        let statement = Statement {
+            source_info: source_info,
+            kind: StatementKind::Assign(
+                ref_loc.clone(),
+                Rvalue::Ref(tcx.types.re_erased, BorrowKind::Shared, rcvr_field)
+            )
+        };
+
+        // `let loc = Clone::clone(ref_loc);`
+        block(blocks, statement, TerminatorKind::Call {
+            func,
+            args: vec![Operand::Consume(ref_loc)],
+            destination: Some((loc.clone(), BasicBlock::new(i + 1))),
+            cleanup: None,
+        });
+
+        loc
+    };
+
+    match rcvr_ty.sty {
+        ty::TyArray(ty, len) => {
+            let mut returns = Vec::new();
```

Comment:
I'm worried this might create code explosion for large arrays (think of `<[Box<u8>; 1048576]>::clone`). Can you try to use a loop and let LLVM unroll it? You're late MIR, so I don't think you have to care about pre-initializing the array.